src/script/sign.h

```diff
@@ -68,6 +71,405 @@ struct SignatureData {
+            // unsigned tx flag
+            WriteCompactSize(s, sizeof(PSBT_UNSIGNED_TX_NON_WITNESS_UTXO));
+            s << PSBT_UNSIGNED_TX_NON_WITNESS_UTXO;
+
+            // Write serialized tx to a stream
+            WriteCompactSize(s, ::GetSerializeSize(tx, SER_NETWORK, PROTOCOL_VERSION));
+            s << tx;
+        }
+
+        // Write redeem scripts and witness scripts
+        for (auto& entry : redeem_scripts) {
+            s << SerializeToVector(PSBT_REDEEMSCRIPT_WITNESS_UTXO, entry.first);
+            s << entry.second;
+        }
+        for (auto& entry : witness_scripts) {
+            s << SerializeToVector(PSBT_WITNESSSCRIPT_PARTIAL_SIG, entry.first);
+            s << entry.second;
+        }
+        // Write any hd keypaths
+        for (auto keypath_pair : hd_keypaths) {
+            s << SerializeToVector(PSBT_BIP32_KEYPATH_SIGHASH, Span<const unsigned char>(keypath_pair.first.begin(), keypath_pair.first.size()));
+            WriteCompactSize(s, keypath_pair.second.size() * sizeof(uint32_t));
+            for (auto& path : keypath_pair.second) {
+                s << path;
+            }
+        }
+
+        // Write the number of inputs
+        if (num_ins > 0) {
+            s << PSBT_NUM_IN_VIN;
+            WriteCompactSize(s, GetSizeOfCompactSize(num_ins));
+            WriteCompactSize(s, num_ins);
+        }
+
+        // Write the unknown things
+        for(auto& entry : unknown) {
+            s << entry.first;
+            s << entry.second;
+        }
+
+        // Separator
+        s << PSBT_SEPARATOR;
+
+        for (unsigned int i = 0; i < tx.vin.size(); i++) {
+            CTxIn in = tx.vin[i];
+            PartiallySignedInput psbt_in = inputs[i];
+            if (in.scriptSig.empty() && in.scriptWitness.IsNull()) {
+                // Write the utxo
+                // If there is a non-witness utxo, then don't add the witness one.
+                if (psbt_in.non_witness_utxo) {
+                    WriteCompactSize(s, sizeof(PSBT_UNSIGNED_TX_NON_WITNESS_UTXO));
+                    s << PSBT_UNSIGNED_TX_NON_WITNESS_UTXO;
+                    WriteCompactSize(s, ::GetSerializeSize(psbt_in.non_witness_utxo, SER_NETWORK, PROTOCOL_VERSION));
+                    s << psbt_in.non_witness_utxo;
+                } else if (!psbt_in.witness_utxo.IsNull()) {
+                    WriteCompactSize(s, sizeof(PSBT_REDEEMSCRIPT_WITNESS_UTXO));
+                    s << PSBT_REDEEMSCRIPT_WITNESS_UTXO;
+                    WriteCompactSize(s, ::GetSerializeSize(psbt_in.witness_utxo, SER_NETWORK, PROTOCOL_VERSION));
+                    s << psbt_in.witness_utxo;
+                }
+
+                // Write any partial signatures
+                for (auto sig_pair : psbt_in.partial_sigs) {
+                    s << SerializeToVector(PSBT_WITNESSSCRIPT_PARTIAL_SIG, Span<const unsigned char>(sig_pair.first.begin(), sig_pair.first.size()));
+                    WriteCompactSize(s, sig_pair.second.size());
```

Comment:
This and the line below can be written as 's << sig_pair.second`, I think - it's just an unsigned char vector.